<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цели-исследования" data-toc-modified-id="Цели-исследования-1"><span class="toc-item-num">1&nbsp;&nbsp;</span><strong>Цели исследования</strong></a></span><ul class="toc-item"><li><span><a href="#Исследование-таблицы----вывод-первых-строк" data-toc-modified-id="Исследование-таблицы----вывод-первых-строк-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Исследование таблицы  - вывод первых строк</a></span></li></ul></li><li><span><a href="#Cколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Cколько-книг-вышло-после-1-января-2000-года-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Cколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзоров-и-средня-оценка--каждой-книги" data-toc-modified-id="Количество-обзоров-и-средня-оценка--каждой-книги-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Количество обзоров и средня оценка  каждой книги</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Автор-с-самой-высокой-средней-оценкой-книги" data-toc-modified-id="Автор-с-самой-высокой-средней-оценкой-книги-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Автор с самой высокой средней оценкой книги</a></span></li><li><span><a href="#Среднее-количество-обзоров-от-пользователей" data-toc-modified-id="Среднее-количество-обзоров-от-пользователей-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Среднее количество обзоров от пользователей</a></span></li></ul></div>

# Анализ базы данных книжного магазина

## **Цели исследования** 

**Задача — проанализировать базу данных, ответить на вопросы**


**Данные  — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Исследование таблицы  - вывод первых строк

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
query = ''' SELECT *
            FROM books
            LIMIT 5;
         '''
books = pd.io.sql.read_sql(query, con = engine)
books

book_id  author_id                                              title   
0        1        546                                       'Salem's Lot  \
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
query = ''' SELECT *
            FROM authors
            LIMIT 5;
         '''
authors = pd.io.sql.read_sql(query, con = engine)
authors

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
query = ''' SELECT *
            FROM publishers
            LIMIT 5;
         '''
publishers = pd.io.sql.read_sql(query, con = engine)
publishers

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
query = ''' SELECT *
            FROM ratings
            LIMIT 5;
         '''
ratings = pd.io.sql.read_sql(query, con = engine)
ratings

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
query = ''' SELECT *
            FROM reviews
            LIMIT 5;
         '''
reviews = pd.io.sql.read_sql(query, con = engine)
reviews

review_id  book_id       username   
0          1        1   brandtandrea  \
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div style="border:solid blue 2px; padding: 20px">
В базе 5 таблиц в них — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг

## Cколько книг вышло после 1 января 2000 года

In [7]:
query = ''' SELECT COUNT(*) 
            FROM books
            WHERE publication_date > '2000-1-1';
         '''
book_count = pd.io.sql.read_sql(query, con = engine)
book_count

count
0    819

<div style="border:solid blue 2px; padding: 20px">
После первого января вышло 819 книг

## Количество обзоров и средня оценка  каждой книги 

In [12]:
query = ''' SELECT books.title,
                   a.review,
                   b.avg_rating
            FROM books 
            LEFT JOIN (SELECT book_id,
                       AVG(rating) AS avg_rating
                       FROM ratings
                       GROUP BY book_id) AS b ON b.book_id = books.book_id
            LEFT JOIN (SELECT book_id,
                       COUNT(review_id) AS review
                       FROM reviews
                       GROUP BY book_id) AS a ON a.book_id = books.book_id
        '''
books1 = pd.io.sql.read_sql(query, con = engine)
books1

title  review  avg_rating
0            The Body in the Library (Miss Marple  #3)     2.0    4.500000
1                                            Galápagos     2.0    4.500000
2                             A Tree Grows in Brooklyn     5.0    4.250000
3    Undaunted Courage: The Pioneering First Missio...     2.0    4.000000
4                                          The Prophet     4.0    4.285714
..                                                 ...     ...         ...
995                                Alice in Wonderland     4.0    4.230769
996          A Woman of Substance (Emma Harte Saga #1)     2.0    5.000000
997                                          Christine     3.0    3.428571
998  The Magicians' Guild (Black Magician Trilogy  #1)     2.0    3.500000
999                           The Plot Against America     2.0    3.000000

[1000 rows x 3 columns]

<div style="border:solid blue 2px; padding: 20px">
Посчитано количество обзоров и средняя оценка для каждой книги

## Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query = ''' SELECT a.publisher,
                   a.n_books
            FROM (SELECT COUNT(books.book_id) AS n_books,
                         publishers.publisher AS publisher 
                         FROM  books 
                  LEFT JOIN publishers ON books.publisher_id=publishers.publisher_id
                            WHERE num_pages>50
                            GROUP BY books.publisher_id,
                            publishers.publisher
                            ORDER BY n_books DESC) AS a
        '''
publishers1 = pd.io.sql.read_sql(query, con = engine)
publishers1.head(1)

publisher  n_books
0  Penguin Books       42

<div style="border:solid blue 2px; padding: 20px">
Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books

## Автор с самой высокой средней оценкой книги

In [10]:
query = ''' SELECT a.author,
                   avg(a.avg) AS rating
            FROM (SELECT authors.author, 
                         AVG(ratings.rating) AS avg
                         FROM books
                  LEFT JOIN authors ON books.author_id = authors.author_id
                  LEFT JOIN ratings ON books.book_id = ratings.book_id
                         GROUP BY books.book_id, authors.author_id
                         HAVING COUNT(ratings.rating_id) >=50 ) AS a 
            GROUP BY a.author
            ORDER BY rating DESC
        '''
authors1 = pd.io.sql.read_sql(query, con = engine)
authors1.head(1)

author    rating
0  J.K. Rowling/Mary GrandPré  4.283844

<div style="border:solid blue 2px; padding: 20px">
Автор с самой высокой средней оценкой книги - это J.K. Rowling/Mary GrandPré

## Среднее количество обзоров от пользователей

In [11]:
query = ''' SELECT AVG(a.b)
            FROM (SELECT username,
                  COUNT(review_id) AS b
                        FROM reviews
                        WHERE username IN ( SELECT username
                                            FROM ratings
                                            GROUP BY username
                                            HAVING COUNT(rating_id) > 50)
                   GROUP BY username) AS a
        '''
review1 = pd.io.sql.read_sql(query, con = engine)
review1

avg
0  24.333333

<div style="border:solid blue 2px; padding: 20px">
cреднее количество обзоров от пользователей - 24.3